Enh arrow json extension

asharmalik19 opened the pull request:
- [x] closes #60958 
- [x] [Tests added and passed]
- [x] All [code checks passed]
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

asharmalik19 committed:
WIP: Support PyArrow JSON type in ArrowDtype.type

asharmalik19 committed:
ENH: Added support for PyArrow JSON type in ArrowDtype.type

asharmalik19 committed:
ENH: Support PyArrow JSON type in ArrowDtype.type

Added support for PyArrow's JSON extension type in ArrowDtype.type by mapping JsonType to str. Fixes #60958.

mroeschke commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -2267,6 +2267,8 @@ def type(self):
             return type(pa_type)
         elif isinstance(pa_type, pa.ExtensionType):
             return type(self)(pa_type.storage_type).type
+        elif isinstance(pa_type, pa.JsonType):
+            return str
```
```suggestion
        elif isinstance(pa_type, pa.BaseExtensionType):
            return type(self)(pa_type.storage_type).type
```

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -65,6 +65,7 @@ Other enhancements
 - :class:`Rolling` and :class:`Expanding` now support ``nunique`` (:issue:`26958`)
 - :class:`Rolling` and :class:`Expanding` now support aggregations ``first`` and ``last`` (:issue:`33155`)
 - :func:`read_parquet` accepts ``to_pandas_kwargs`` which are forwarded to :meth:`pyarrow.Table.to_pandas` which enables passing additional keywords to customize the conversion to pandas, such as ``maps_as_pydicts`` to read the Parquet map data type as python dictionaries (:issue:`56842`)
+- :meth: ``ArrowDtype.type`` now supports the pyarrow json data type (:issue:`60958`)
```
```suggestion
- :class:`ArrowDtype` now supports ``pyarrowJsonType`` (:issue:`60958`)
```

asharmalik19 committed:
ENH: Support for PyArrow JSON type in ArrowDtype.type

Improved extension type handling by using BaseExtensionType for consistent storage type resolution across all PyArrow extension types, including JSON.

Fixes #60958

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -61,6 +61,7 @@ Other enhancements
 - :meth:`Series.cummin` and :meth:`Series.cummax` now supports :class:`CategoricalDtype` (:issue:`52335`)
 - :meth:`Series.plot` now correctly handle the ``ylabel`` parameter for pie charts, allowing for explicit control over the y-axis label (:issue:`58239`)
 - :meth:`DataFrame.plot.scatter` argument ``c`` now accepts a column of strings, where rows with the same string are colored identically (:issue:`16827` and :issue:`16485`)
+- :class:`ArrowDtype` now supports ``pyarrowJsonType`` (:issue:`60958`)
```
```suggestion
- :class:`ArrowDtype` now supports ``pyarrow.JsonType`` (:issue:`60958`)
```

mroeschke commented on pandas/tests/extension/test_arrow.py:
```diff
@@ -3553,3 +3553,10 @@ def test_categorical_from_arrow_dictionary():
     tm.assert_series_equal(result, expected)
+
+
+def test_arrow_json_type():
```
You'll need to `@pytest.mark.skipif(pa_version_under19p0, reason=...)` as `pa.json_` is new as of Pyarrow 19

mroeschke commented:
Thanks @asharmalik19

asharmalik19 commented:
Thanks @mroeschke for such quick feedback

tswast commented:
Thank you so much for your help, @asharmalik19 and @mroeschke !
